File: panther_battery/src/roboteq_republisher_node.py
#!/usr/bin/python3

from threading import Lock
from typing import Optional

import rospy

from sensor_msgs.msg import BatteryState

from panther_msgs.msg import DriverState


class RoboteqRepublisherNode:
    V_BAT_FATAL_MIN = 27.0
    V_BAT_FATAL_MAX = 43.0
    V_BAT_FULL = 41.4
    V_BAT_MIN = 32.0

    def __init__(self, name: str) -> None:
        rospy.init_node(name, anonymous=False)

        self._lock = Lock()

        self._battery_voltage: Optional[float] = None
        self._battery_current: Optional[float] = None
        self._battery_voltage_mean = 37.0

        self._volt_mean_length = 10
        self._battery_voltage_hist = [37.0] * self._volt_mean_length

        self._battery_timeout = 1.0
        self._last_battery_info_time = rospy.get_time()

        # -------------------------------
        #   Subscribers
        # -------------------------------

        self._motor_controllers_state_sub = rospy.Subscriber(
            'driver/motor_controllers_state', DriverState, self._motor_controllers_state_cb
        )

        # -------------------------------
        #   Publishers
        # -------------------------------

        self._battery_pub = rospy.Publisher('battery', BatteryState, queue_size=1)

        # -------------------------------
        #   Timers
        # -------------------------------

        # Running at 10 HZ
        self._battery_pub_timer = rospy.Timer(rospy.Duration(0.1), self._battery_pub_timer_cb)

        rospy.loginfo(f'[{rospy.get_name()}] Node started')

    def _motor_controllers_state_cb(self, msg: DriverState) -> None:
        with self._lock:
            new_voltage = (msg.front.voltage + msg.rear.voltage) / 2.0

            self._last_battery_info_time = rospy.get_time()
            self._battery_voltage = new_voltage
            self._battery_current = msg.front.current + msg.rear.current
            self._update_volt_mean(new_voltage)

    def _battery_pub_timer_cb(self, *args) -> None:
        with self._lock:
            battery_msg = BatteryState()
            battery_msg.header.stamp = rospy.Time.now()
            battery_msg.capacity = 20.0
            battery_msg.design_capacity = 20.0
            battery_msg.power_supply_technology = BatteryState.POWER_SUPPLY_TECHNOLOGY_LIPO

            if (
                self._battery_voltage == None
                or self._battery_current == None
                or rospy.get_time() - self._last_battery_info_time > self._battery_timeout
            ):
                battery_msg.power_supply_status = BatteryState.POWER_SUPPLY_STATUS_UNKNOWN
            else:
                battery_msg.voltage = self._battery_voltage
                battery_msg.temperature = float('nan')
                battery_msg.current = self._battery_current
                battery_msg.percentage = self._clamp(
                    (battery_msg.voltage - self.V_BAT_MIN) / (self.V_BAT_FULL - self.V_BAT_MIN),
                    0.0,
                    1.0,
                )
                battery_msg.charge = battery_msg.percentage * battery_msg.design_capacity
                battery_msg.present = True

                battery_msg.power_supply_status = BatteryState.POWER_SUPPLY_STATUS_DISCHARGING

                # check battery health
                error_msg = None
                if self._battery_voltage < self.V_BAT_FATAL_MIN:
                    battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_DEAD
                    error_msg = 'Battery voltage is critically low!'
                elif self._battery_voltage > self.V_BAT_FATAL_MAX:
                    battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_OVERVOLTAGE
                    error_msg = 'Battery overvoltage!'
                else:
                    battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_GOOD

                if error_msg is not None:
                    rospy.logerr_throttle_identical(10.0, f'[{rospy.get_name()}] {error_msg}')

            self._battery_pub.publish(battery_msg)

    def _update_volt_mean(self, new_val: float) -> float:
        # Updates the average by adding the newest and removing the oldest component of mean value,
        # in order to avoid recalculating the entire sum every time.
        self._battery_voltage_mean += (
            new_val - self._battery_voltage_hist[0]
        ) / self._volt_mean_length

        self._battery_voltage_hist.pop(0)
        self._battery_voltage_hist.append(new_val)

    @staticmethod
    def _clamp(value, min_value, max_value):
        return max(min(value, max_value), min_value)


def main():
    roboteq_republisher_node = RoboteqRepublisherNode('roboteq_republisher_node')
    rospy.spin()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: panther_battery/src/roboteq_republisher_node.cpp
#include <panther_battery/roboteq_republisher_node.hpp>

#include <chrono>
#include <functional>
#include <limits>
#include <memory>

#include <rclcpp/rclcpp.hpp>

#include <sensor_msgs/msg/battery_state.hpp>

#include <panther_msgs/msg/driver_state.hpp>

#include <panther_utils/moving_average.hpp>

namespace panther_battery
{
using std::placeholders::_1;

RoboteqRepublisherNode::RoboteqRepublisherNode(
  const std::string & node_name, const rclcpp::NodeOptions & options)
: Node(node_name, options)
{
  this->declare_parameter<float>("battery_timeout", 1.0);
  this->declare_parameter<int>("batery_voltage_window_len", 10);
  this->declare_parameter<int>("batery_current_window_len", 10);

  battery_timeout_ = this->get_parameter("battery_timeout").as_double();
  const auto batery_voltage_window_len = this->get_parameter("batery_voltage_window_len").as_int();
  const auto batery_current_window_len = this->get_parameter("batery_current_window_len").as_int();

  battery_voltage_ma_ = std::make_unique<panther_utils::MovingAverage<double>>(
    batery_voltage_window_len, std::numeric_limits<double>::quiet_NaN());
  battery_current_ma_ = std::make_unique<panther_utils::MovingAverage<double>>(
    batery_current_window_len, std::numeric_limits<double>::quiet_NaN());

  motor_controllers_state_sub_ = this->create_subscription<DriverStateMsg>(
    "driver/motor_controllers_state", 10,
    std::bind(&RoboteqRepublisherNode::MotorControllersStateSubCB, this, _1));

  battery_pub_ = this->create_publisher<BatteryStateMsg>("battery", 10);

  // running at 10 Hz
  battery_pub_timer_ = this->create_wall_timer(
    std::chrono::milliseconds(100), std::bind(&RoboteqRepublisherNode::BatteryPubTimerCB, this));

  RCLCPP_INFO(this->get_logger(), "Node started");
}

void RoboteqRepublisherNode::MotorControllersStateSubCB(const DriverStateMsg & msg)
{
  if (msg.front.fault_flag.can_net_err || msg.rear.fault_flag.can_net_err) {
    return;
  }

  last_battery_info_time_ = this->get_clock()->now();
  battery_voltage_ma_->Roll((msg.front.voltage + msg.rear.voltage) / 2.0);
  battery_current_ma_->Roll(msg.front.current + msg.rear.current);
}

void RoboteqRepublisherNode::BatteryPubTimerCB()
{
  auto battery_msg = BatteryStateMsg();
  battery_msg.header.stamp = this->get_clock()->now();
  battery_msg.capacity = bat_capacity_;
  battery_msg.design_capacity = bat_designed_capacity_;
  battery_msg.temperature = std::numeric_limits<float>::quiet_NaN();
  battery_msg.power_supply_technology = BatteryStateMsg::POWER_SUPPLY_TECHNOLOGY_LIPO;
  battery_msg.present = true;
  battery_msg.cell_voltage = std::vector<float>(10, std::numeric_limits<float>::quiet_NaN());
  battery_msg.cell_temperature = std::vector<float>(10, std::numeric_limits<float>::quiet_NaN());
  battery_msg.location = location_;

  auto battery_voltage = battery_voltage_ma_->GetAverage();
  auto battery_current = battery_current_ma_->GetAverage();

  if (std::isnan(battery_voltage) || std::isnan(battery_current)) {
    battery_msg.power_supply_status = BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN;
    battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN;
  } else if (
    (this->get_clock()->now() - last_battery_info_time_) >
    rclcpp::Duration::from_seconds(battery_timeout_)) {
    battery_voltage_ma_->Reset();
    battery_current_ma_->Reset();
    battery_voltage = battery_voltage_ma_->GetAverage();
    battery_current = battery_current_ma_->GetAverage();
    battery_msg.power_supply_status = BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN;
    battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN;
  } else {
    battery_msg.power_supply_status = BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING;

    // check battery health
    if (battery_voltage < v_bat_fatal_min_) {
      battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_DEAD;
      RCLCPP_ERROR_THROTTLE(
        this->get_logger(), *this->get_clock(), 10000, "Battery voltage is critically low!");
    } else if (battery_voltage > v_bat_fatal_max_) {
      battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_OVERVOLTAGE;
      RCLCPP_ERROR_THROTTLE(this->get_logger(), *this->get_clock(), 10000, "Battery overvoltage!");
    } else {
      battery_msg.power_supply_health = BatteryStateMsg::POWER_SUPPLY_HEALTH_GOOD;
    }
  }

  battery_msg.voltage = battery_voltage;
  battery_msg.current = battery_current;
  battery_msg.percentage =
    std::clamp((battery_voltage - v_bat_min_) / (v_bat_full_ - v_bat_min_), 0.0, 1.0);
  battery_msg.charge = battery_msg.percentage * battery_msg.design_capacity;

  battery_pub_->publish(battery_msg);
}

}  // namespace panther_battery